### PR TITLE
Video Recording: Add initial implementation

### LINKED ISSUE #10958: Video Capture: Initial implementation
## Feature Description

As per #10930 we want to support video/audio capture.

This ticket is for adding the initial plumbing for this feature, based on the existing proof of concept.

We can then build off of this initial implementation.

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

As per #10958 / #10930, we want to add support for webcam/microphone recording.

## Summary

This adds a the foundation for the following work to enable video/audio/image recording in the editor.

This is very rough around the edges. There's absolutely no UX for this so far and some known bugs.

But it gets the job done and unblocks all future work in this area.

## Relevant Technical Choices

* Fixes a bug in `getVideoLength` which returned `NaN` for recorded videos because the video duration was returned as `Infinity`.
* Uses `@wmik/use-media-recorder` instead of implementing all this logic myself. While it's relatively new and marked as a beta, it seems to work quite well. We can revisit this choice later if needed.

## To-do

* Tests
* UX
* Telemetry
* Bugfixes
* ... and lots more

Lots of polish is needed in future PRs. And of course actual UX.

## User-facing changes

Once the experiment is enabled, there is a new recording button to open the dialog for webcam recording.

https://user-images.githubusercontent.com/841956/158474813-e816259e-1a1e-4d73-954c-11222265e183.mov

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No, something for future PRs

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #10958

### CODE REVIEW

File: packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+} from '@googleforcreators/react';
+import styled from 'styled-components';
+import PropTypes from 'prop-types';
+import useMediaRecorder from '@wmik/use-media-recorder';
+import { blobToFile } from '@googleforcreators/media';
+
+/**
+ * Internal dependencies
+ */
+import { v4 as uuidv4 } from 'uuid';
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js]: Move to External dependencies

File: packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+`;
+
+function Modal({ isOpen, onClose }) {
+  const videoRef = useRef();
+  const [file, setFile] = useState(null);
+  const [enableVideo, setEnableVideo] = useState(true);
+  const [enableAudio, setEnableAudio] = useState(true);
+
+  const onStop = useCallback((blob) => {
+    const f = blobToFile(blob, 'test-recording.webm', 'video/webm');
```
Comment (spacedmonkey) [inline packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js:67]: Why are we using webm here? We mostly use mp4s in this editor. Also, can be name the file as the timestamp or something.

**Comment by spacedmonkey** [comment]
https://user-images.githubusercontent.com/237508/159749873-34e6b8dd-914f-4103-9a1e-0ab065771473.mov

When I open the dialog for the first time, I don't see a dialog asking me to allow for camera use. Is this a mistake?

**Comment by spacedmonkey** [comment]
Is there a possibility to add an option to select which camera to use. Like in google hangout?
Screenshot 2022-03-23 at 16 38 28

**Comment by spacedmonkey** [comment]
What happens if the user do not have a camera at all? Some error handling or hide the camera button?

File: packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js (hunk 2 above)
Comment (swissspidy) [reply packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js:67]: For this initial implementation I stuck with the mime type Chrome uses for media recording, which is webm. So it returns a webm, which is why this is a webm.

In future versions this needs to be dynamic.

Determining the file name is something I'll bring up with UX as well.

**Comment by swissspidy** [comment]
> When I open the dialog for the first time, I don't see a dialog asking me to allow for camera use. Is this a mistake?

Right now the dialog opens when you first click on the "Start recording" button.

This is likely to change once we have UX.

> Is there a possibility to add an option to select which camera to use. Like in google hangout?

Yes this is technically possible and something that will be considered for UX, but out of scope for this PR.

> What happens if the user do not have a camera at all? Some error handling or hide the camera button?

This also will be considered for UX, but out of scope for this PR.

**Comment by spacedmonkey** [comment]
Works well when video transcoding is disabled 👍

**Comment by spacedmonkey** [comment]
One thing I found odd, when I open the dialog, record a video and then inserted it. When I opened dialog again, the prevous video I record was still there. I think the dialog should reset after is closed. If I want to see the video I recorded before, I see that in my media library.

File: packages/media/src/getVideoLength.js [hunk 3]
```diff
@@ -26,7 +26,8 @@ import getVideoLengthDisplay from './getVideoLengthDisplay';
  * @return {Object<{length: number, lengthFormatted: string}>} Video length information.
  */
 function getVideoLength(video) {
-  const duration = !isNaN(video.duration) ? video.duration : 0;
+  const isNumber = !isNaN(video.duration) && Infinity !== video.duration;
```
Comment (spacedmonkey) [inline packages/media/src/getVideoLength.js:31]: I would add a comment to explain why this is needed.

File: packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js (hunk 2 above)
Comment (spacedmonkey) [reply packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js:67]: Maybe we should add a todo here, to change this to an different format.

**Comment by spacedmonkey** [comment]
This is GREAT for a first pass. My only feedback, is that on close, the video / dialog to reset. Other than this, it is basically ready to go behind a feature.

**Review by merapi — APPROVED** [review]
Good PoC, just one thing that was already mentioned: changing the recording device(s), because for me OBS is first on the list and I need to go into Chrome settings to change that - some people will struggle here.

image

**Comment by swissspidy** [comment]
> I think the dialog should reset after is closed.

Yeah that's a known issue that will need to get addressed. Likely will get fixed automatically once we change the UX flow.

**Comment by timarney** [comment]
As per other comments I think this is great as a POC.

Just to capture first thoughts as a user...

> For the initial screen the "black box" pre giving the camera permission made me think something was broken.   Given the Camera and Mic are already toggled to the on state. Might suggest moving that prompt to when you first press "the camera" button.  

Screen Shot 2022-03-25 at 4 52 30 AM

**Review by timarney — APPROVED** [review]

File: packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js (hunk 2 above)
Comment (swissspidy) [reply packages/story-editor/src/components/library/panes/media/local/mediaRecording/modal.js:67]: No need for a todo. I'll create tickets for this.

**Pull request merged**